src/_pytest/tmpdir.py

```diff
@@ -2,21 +2,240 @@
+
+
+def try_cleanup(path, consider_lock_dead_after):
+    if ensure_deletable(path, consider_lock_dead_after):
+        delete_a_numbered_dir(path)
+
+
+def cleanup_candidates(root, prefix, keep):
+    max_existing = _max(map(parse_num, find_suffixes(root, prefix)), -1)
+    max_delete = max_existing - keep
+    paths = find_prefixed(root, prefix)
+    paths, paths2 = itertools.tee(paths)
+    numbers = map(parse_num, extract_suffixes(paths2, prefix))
+    for path, number in zip(paths, numbers):
+        if number <= max_delete:
+            yield path
+
+
+def cleanup_numbered_dir(root, prefix, keep, consider_lock_dead_after):
+    for path in cleanup_candidates(root, prefix, keep):
+        try_cleanup(path, consider_lock_dead_after)
+    known_garbage = list(root.glob("garbage-*"))
+    for path in known_garbage:
+        try_cleanup(path, consider_lock_dead_after)
 
 
+def make_numbered_dir_with_cleanup(root, prefix, keep, lock_timeout):
+    for i in range(10):
+        try:
+            p = make_numbered_dir(root, prefix)
+            lock_path = create_cleanup_lock(p)
+            register_cleanup_lock_removal(lock_path)
+        except Exception:
+            raise
+        else:
+            consider_lock_dead_after = p.stat().st_mtime + lock_timeout
+            cleanup_numbered_dir(
+                root=root,
+                prefix=prefix,
+                keep=keep,
+                consider_lock_dead_after=consider_lock_dead_after,
+            )
+            return p
+
+
+@attr.s
+class TempPathFactory(object):
+    """docstring for ClassName"""
+
+    given_basetemp = attr.ib()
+    trace = attr.ib()
+    _basetemp = attr.ib(default=None)
+
+    @classmethod
+    def from_config(cls, config):
+        return cls(
+            given_basetemp=config.option.basetemp, trace=config.trace.get("tmpdir")
+        )
+
+    def mktemp(self, basename, numbered=True):
+        if not numbered:
+            p = self.getbasetemp().joinpath(basename)
+            p.mkdir()
+        else:
+            p = make_numbered_dir(root=self.getbasetemp(), prefix=basename)
+            self.trace("mktemp", p)
+        return p
+
+    def getbasetemp(self):
+        """ return base temporary directory. """
+        if self._basetemp is None:
+            if self.given_basetemp is not None:
+                basetemp = Path(self.given_basetemp)
+                if basetemp.exists():
+                    shutil.rmtree(str(basetemp))
+                basetemp.mkdir()
+            else:
+                temproot = Path(tempfile.gettempdir())
+                user = get_user() or "unknown"
+                # use a sub-directory in the temproot to speed-up
+                # make_numbered_dir() call
+                rootdir = temproot.joinpath("pytest-of-{}".format(user))
+                rootdir.mkdir(exist_ok=True)
+                basetemp = make_numbered_dir_with_cleanup(
+                    prefix="pytest-", root=rootdir, keep=3, lock_timeout=10000
```

Comment:
10000 seconds is 2.7 hours, the original make_numbered_dir from `py` [uses 172800](https://github.com/pytest-dev/py/blob/796d5a0d92f6eef6f5e28dfffa3a613c73a77b22/py/_path/local.py#L803). Is this intentional?

Also it helps to write `60 * 60 * 3` instead of `10800  # 3 hours` IMHO.